source/lac/dynamic_sparsity_pattern.cc

```diff
@@ -321,10 +325,114 @@ DynamicSparsityPattern::reinit(const size_type m,
 
 
 
+#ifdef DEAL_II_WITH_MPI
+void
+DynamicSparsityPattern::reinit(const size_type m,
+                               const size_type n,
+                               const IndexSet &owned_rows_,
+                               const MPI_Comm  mpi_communicator_,
+                               const IndexSet &rowset)
+{
+  reinit(m, n, rowset);
+  owned_rows       = owned_rows_;
+  mpi_communicator = mpi_communicator_;
+}
+#endif
+
+
+
 void
 DynamicSparsityPattern::compress()
-{}
+{
+#ifdef DEAL_II_WITH_MPI
+  // don't do anything if we did not call reinit() with MPI_Comm and IndexSet
+  if (owned_rows.size() == 0)
+    return;
+
+  const unsigned int myid = Utilities::MPI::this_mpi_process(mpi_communicator);
+
+  Assert(owned_rows.is_ascending_and_one_to_one(mpi_communicator),
+         ExcNotImplemented());
+  const auto b = owned_rows.nth_index_in_set(0);
+  const auto e = owned_rows.nth_index_in_set(owned_rows.n_elements() - 1) + 1;
```

Comment:
I think you might want to write a test in which a processor doesn't actually owns any rows and make sure that the code here works (which I don't think it does, currently, because `nth_index_in_set(0)` will fail). This is a common case if you start with a single cell and more than one processor.